File: Translation_No_1/2-R/lec2.r
#################
# Lecture 2 live
#################

# Topics of this lecture
# 1) Juggling with dataframes
# 2) Reading data from a website
# 3) Package
# 4) Loops
# 5) Drawing random numbers from a particular
#   distribution (if we get there)

# Install XML package
# Do this only once per machine!
# It's like an app
# install.packages("XML")

# Call the package/library
# Do this in every session
rm(list = ls())
library("XML")

shopList_0 = readHTMLTable("http://www.einkaufszettel.de/einkaufslisten/einkaufszettel-vorlage")
class(shopList_0)
shopList_0
## [1] "list"
# This is a list, new data type
length(shopList_0)
## [1] 2
# Select second list item
shopList = shopList_0[[2]]
class(shopList)
## [1] "data.frame"
class(shopList$V1)
## [1] "factor"
# The type of this variable is factor
# and we don't like that, for what we want to do
# We want "character"
shopList[] = lapply(shopList, as.character)
class(shopList$V1)
## [1] "character"
test1 = shopList[   1 , #rows 
                      ] #columns
                        # empty means ALL!
test1 = shopList[1, ]
class(test1)
## [1] "data.frame"
test2 = shopList[ , 1]
class(test2)
## [1] "character"
test3 = shopList[ , c(1,2)]

# delete row 1 and 18 from shopList
shopList = shopList[ -c(1,18) ,  ]

#columns to select
sel = seq(1, 9, by=2)

shopList1 = shopList[ , sel]

rm(list = ls(pattern = "^test"))

# Tests for getting all shopping items in one column
test1 = c(shopList$V1,
          shopList$V3,
          shopList$V5,
          shopList$V7,
          shopList$V9
          )


test1 = as.data.frame(test1)

test2 = c() #initialization
for (i in 1:ncol(shopList)){
  # Here is our algorithm
  test2 = c(test2, shopList[ , i])
}
test2 = as.data.frame(test2)

shopList = test2
class(shopList)
## [1] "data.frame"
# Change name of the single variable in shopList
names(shopList) = "item"
class(shopList$item)
## [1] "factor"
#Convert this to character
shopList[] = lapply(shopList, as.character)
class(shopList$item)
shopList[,1]
## [1] "character"
sink("inspectWhatWeAreDoing.txt", append=FALSE)
shopList$item == ""
##   [1] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
##  [12]  TRUE  TRUE  TRUE  TRUE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE
##  [23] FALSE FALSE FALSE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE
##  [34]  TRUE  TRUE  TRUE  TRUE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE
##  [45] FALSE FALSE FALSE FALSE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE FALSE
##  [56] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
##  [67] FALSE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE  TRUE  TRUE FALSE
##  [78] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
##  [89] FALSE  TRUE  TRUE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
## [100] FALSE FALSE FALSE FALSE FALSE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE
## [111]  TRUE  TRUE  TRUE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
## [122] FALSE FALSE FALSE FALSE FALSE FALSE  TRUE  TRUE  TRUE FALSE FALSE
## [133] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
## [144] FALSE FALSE FALSE FALSE FALSE  TRUE  TRUE  TRUE  TRUE FALSE FALSE
## [155] FALSE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE  TRUE  TRUE  TRUE
## [166]  TRUE  TRUE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
## [177] FALSE FALSE  TRUE FALSE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE
## [188]  TRUE  TRUE  TRUE
which(shopList$item == "")
##  [1]  12  13  14  15  16  27  28  29  30  31  32  33  34  35  36  37  38
## [18]  50  51  52  53  54  74  75  76  90  91  92 106 107 108 109 110 111
## [35] 112 113 114 128 129 130 149 150 151 152 162 163 164 165 166 167 168
## [52] 179 182 183 184 185 186 187 188 189 190
sink()
file.show("inspectWhatWeAreDoing.txt")

# In class, we got until here. 

# Now we still want to drop the empty rows from shopList
# In class, I still typed the following (how commented out with hashtags)

# shopList = shopList[-which(shopList$item == "") , ]

# This does not yield the desired result, for somewhat complicated reasons
# Let's not bother right now. 
# Instead, the following works

shopList = as.data.frame(shopList$item[-which(shopList$item == "") ])
names(shopList) = "item"

File: Translation_No_1/1-Python/Lec2.py
#################
# Lecture 2 live
#################

# Topics of this lecture
# 1) Juggling with dataframes
# 2) Reading data from a website
# 3) Package
# 4) Loops
# 5) Drawing random numbers from a particular


# we again first of all import all our packages which we are going to need
import pandas as pd
import numpy as np
import sys
import os

# set our directory where we will get our files from and save new ones to
os.chdir("/Users/dominiquepaul/xJob/1-DataWithPythonCourse")

# we can also read in data from websites
shop_list_0 = pd.read_html("http://www.einkaufszettel.de/einkaufslisten/einkaufszettel-vorlage")

# This is a list
print(type(shop_list_0))
print(len(shop_list_0))
# in it there are various data frames
print(type(shop_list_0[1]))

# Select second list item
shop_list = shop_list_0[1]
print(type(shop_list))


########################################
# Selecting items from a data frame
########################################

# similar to R, we can grab items from a df via their position
# however, different than in R, items are indexed starting with 0
# so if you want to get the first row or column, you have to use '0' 

# we select the second row and all columns here
# even if we are not selecting a row or column, we have to use ':' as a placeholder 
# as the function wont work otherwise

test1 = shop_list.ix[1,  #rows
 :] #columns
#we could use squared brackets after our selection of the row/column to select the n-th item of our selection if relevant

# selecting all rows and the second column
test2 = shop_list.ix[:,1]
print(type(test2))

# selecting the first item from all rows and the second column 
test3 = shop_list.ix[:,1][0]
print(test3)

# selecting the first two rows of the data frame and all rows
test4 = shop_list.ix[0:2,:]
print(test4.head())

# note that the notation is start:stop, thus we start with row number 0,
# but we stop BEFORE row 3 (which is 2 in the index). We thus get rows 0 and 1

# deleting rows and columns from the list:
shop_list1 = shop_list.drop(0, # indicates the row or column to be dropped (remember indexing begins with 0 )
	axis = 0) # axis indicates whether we want to drop a row or column (0 = row, 1 = column)
print(shop_list1.head(2))

# dropping multiple rows at the same time
shop_list = shop_list.drop([0,2], axis = 0)
print(shop_list.head(2))

# creating an array with the columns we want to select
sequence = np.linspace(0, #starting point
	8, # ending point
	5) # amount of evenly spaced numbers to be created
print(sequence)

# we only select the columns which are names
shop_list1 = shop_list.ix[:,sequence]

# lets delete all 'test' variables as we dont need them anymore 
# and in order to avoid later confusion
# dir() gives us a list of all variables existing in python
# via the for loop we can iterate through all of then and check whether
# the begin with test, if so we delete them
for name in dir():
    if name.startswith('test'):
        del globals()[name]

# Tests for getting all shopping items in one column
test1 = pd.concat([shop_list.ix[:,0],
	shop_list.ix[:,2],
	shop_list.ix[:,4], 
	shop_list.ix[:,6],
	shop_list.ix[:,8]])

# convert the file from a series to a dataframe
test1 = pd.DataFrame(test1)

# initialize an empty series (similar to array)
test2 = pd.Series([])

# range creates an array from 0 to the amount of columns through which we can iterate
for i in range(0, len(shop_list.columns)):
	# this combines the two elements to one
	test2 = pd.concat([test2, shop_list.ix[:,i]])

test2 = pd.DataFrame(test2)

shop_list = test2
print(type(shop_list))
print(shop_list)


# we see that the index is not constant, lets change this: 
# resets the index, this would push in the old one as a columns so we delete it with the drop command
shop_list.reset_index(drop = True, # by resetting the index the 'old index' would be pushed in as a new index, this command stops this from happening
 inplace = True) # allows the command to be used on the df itself, otherwise we would have to use df = df.reset_index(...)

# change column name
shop_list.columns = ["item"]
print(shop_list)


# as we see not all items are strings though
print(type(shop_list.item[12]))
# or by looking at 
dtypeCount = shop_list.item.apply(type).value_counts()
print(dtypeCount)

# lets change this
shop_list['item'] = shop_list['item'].astype('str') 

###################################################
# printing our output to a seperate text file
###################################################

# save old system output (the place where python directs its output) 
# so we can change back later if we wish
orig_stdout = sys.stdout
# create file where we want to save the output
file = open('inspectWhatWeAreDoing.txt', 'w')
# change output to normal 
sys.stdout = file

# compares to see which rows are empty
print(shop_list.ix[:,'item'] == "nan")

# prints the row numbers where the rows are empty
print(shop_list.loc[shop_list['item'] == "nan"].index.tolist())

# revert output to original location
sys.stdout = orig_stdout
# save file
file.close()


# Now we still want to drop the empty rows from shopList

list_to_delete = shop_list.loc[shop_list['item'] == "nan"].index.tolist()
shop_list1 = shop_list.drop(list_to_delete, axis = 0)

print(shop_list1)
